File: android/src/main/java/com/example/aliyunplayview/AliyunVodDownloadModule.java
package com.example.aliyunplayview;

import android.util.Log;

import com.aliyun.vodplayer.downloader.AliyunDownloadConfig;
import com.aliyun.vodplayer.downloader.AliyunDownloadInfoListener;
import com.aliyun.vodplayer.downloader.AliyunDownloadManager;
import com.aliyun.vodplayer.downloader.AliyunDownloadMediaInfo;
import com.aliyun.vodplayer.downloader.AliyunRefreshPlayAuthCallback;
import com.aliyun.vodplayer.media.AliyunPlayAuth;
import com.cedarsoftware.util.io.JsonReader;
import com.example.aliyunplayview.util.MapUtil;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.JavaScriptModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class AliyunVodDownloadModule extends ReactContextBaseJavaModule {
    private final String mStoragePath;

    private ReactApplicationContext mContext;
    private AliyunDownloadManager mDownloadManager;
    private Lock mAuthInfoLock = new ReentrantLock();
    private Condition mAuthInfoCond = mAuthInfoLock.newCondition();
    private boolean mAuthInfoSet = false;
    private String mAuthInfoVidId = "";
    private String mAuthInfoAuthStr = "";
    private DeviceEventManagerModule.RCTDeviceEventEmitter mDeviceEmitter;

    private static final String EVENT_UPDATE_AUTH = "AliyunVod.Downloader.UpdateAuth";
    private static final String EVENT_PREPARED = "AliyunVod.Downloader.Prepared";
    private static final String EVENT_START = "AliyunVod.Downloader.Start";
    private static final String EVENT_PROGRESS = "AliyunVod.Downloader.Progress";
    private static final String EVENT_COMPLETE = "AliyunVod.Downloader.Completed";
    private static final String EVENT_STOP = "AliyunVod.Downloader.Stop";
    private static final String EVENT_ERROR = "AliyunVod.Downloader.Error";

    AliyunVodDownloadModule(ReactApplicationContext context) {
        super(context);
        mContext = context;
        mDownloadManager = AliyunDownloadManager.getInstance(context);
        mStoragePath = context.getFilesDir().toString() + "/vodDownload/";
    }

    public String getName() {
        return "AliyunVodDownloadModule";
    }

    @Override
    public void initialize() {
        super.initialize();

        AliyunDownloadConfig config = new AliyunDownloadConfig();
        config.setDownloadDir(mStoragePath);
        config.setSecretImagePath("");
        this.mDownloadManager.setDownloadConfig(config);

        mDeviceEmitter = mContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class);
        this.setDownloadManagerRefreshAuth();
        this.bindDownloadEvents();
    }

    private void setDownloadManagerRefreshAuth() {
        this.mDownloadManager.setRefreshAuthCallBack(new AliyunRefreshPlayAuthCallback() {
            @Override
            public AliyunPlayAuth refreshPlayAuth(String vid, String quality, String format, String title, boolean encrypt) {
                WritableMap args = Arguments.createMap();
                args.putString("vidId", vid);
                args.putString("quality", quality);
                args.putString("format", format);
                args.putString("title", title);
                args.putBoolean("encrypt", encrypt);

                mDeviceEmitter.emit(EVENT_UPDATE_AUTH, args);

                mAuthInfoLock.lock();
                try {
                    while(!mAuthInfoSet) {
                        mAuthInfoCond.await();
                    }
                } catch (InterruptedException e) {
                    Log.e("ReactNative", "refreshPlayAuth: Interrupted" );
                    mAuthInfoSet = false;
                }
                mAuthInfoLock.unlock();

                AliyunPlayAuth.AliyunPlayAuthBuilder playAuthBuilder = new AliyunPlayAuth.AliyunPlayAuthBuilder();
                playAuthBuilder.setVid(mAuthInfoVidId);
                playAuthBuilder.setQuality(quality);
                playAuthBuilder.setPlayAuth(mAuthInfoAuthStr);
                playAuthBuilder.setFormat(format);
                playAuthBuilder.setTitle(title);
                playAuthBuilder.setIsEncripted(encrypt? 1 : 0);

                return playAuthBuilder.build();
            }
        });
    }

    private void bindDownloadEvents() {
        mDownloadManager.setDownloadInfoListener(new AliyunDownloadInfoListener() {
            @Override
            public void onPrepared(List<AliyunDownloadMediaInfo> list) {
                mDeviceEmitter.emit(EVENT_PREPARED, mediaInfosToWritableMap(list));
            }

            @Override
            public void onStart(AliyunDownloadMediaInfo aliyunDownloadMediaInfo) {
                Log.d("ReactNative", "onStart: ");
                mDeviceEmitter.emit(EVENT_START, mediaInfosToWritableMap(aliyunDownloadMediaInfo));
            }

            @Override
            public void onProgress(AliyunDownloadMediaInfo aliyunDownloadMediaInfo, int progress) {
                WritableMap args = Arguments.createMap();
                args.putMap("media", mediaInfosToWritableMap(aliyunDownloadMediaInfo));
                args.putInt("progress", progress);
                mDeviceEmitter.emit(EVENT_PROGRESS, args);
            }

            @Override
            public void onStop(AliyunDownloadMediaInfo aliyunDownloadMediaInfo) {
                mDeviceEmitter.emit(EVENT_STOP, mediaInfosToWritableMap(aliyunDownloadMediaInfo));
            }

            @Override
            public void onCompletion(AliyunDownloadMediaInfo aliyunDownloadMediaInfo) {
                mDeviceEmitter.emit(EVENT_COMPLETE, mediaInfosToWritableMap(aliyunDownloadMediaInfo));
            }

            @Override
            public void onError(AliyunDownloadMediaInfo aliyunDownloadMediaInfo, int i, String s, String s1) {
                WritableMap args = Arguments.createMap();
                args.putMap("media", mediaInfosToWritableMap(aliyunDownloadMediaInfo));
                args.putInt("i", i);
                args.putString("s1", s);
                args.putString("s2", s1);

                mDeviceEmitter.emit(EVENT_ERROR, args);
            }

            @Override
            public void onWait(AliyunDownloadMediaInfo aliyunDownloadMediaInfo) {

            }

            @Override
            public void onM3u8IndexUpdate(AliyunDownloadMediaInfo aliyunDownloadMediaInfo, int i) {

            }
        });
    }

    private WritableMap mediaInfosToWritableMap(List<AliyunDownloadMediaInfo> infos) {
        String json = AliyunDownloadMediaInfo.getJsonFromInfos(infos);
        Map rawMap = JsonReader.jsonToMaps(json);
        return MapUtil.toWritableMap(rawMap);
    }

    private WritableMap mediaInfosToWritableMap(AliyunDownloadMediaInfo infos[]) {
        return this.mediaInfosToWritableMap(Arrays.asList(infos));
    }

    private WritableMap mediaInfosToWritableMap(AliyunDownloadMediaInfo info) {
        return this.mediaInfosToWritableMap(new AliyunDownloadMediaInfo[]{info});
    }

    @ReactMethod
    public void setAuth(String vidId, String authStr) {
        this.mAuthInfoLock.lock();
        this.mAuthInfoVidId = vidId;
        this.mAuthInfoAuthStr = authStr;
        this.mAuthInfoSet = true;
        this.mAuthInfoCond.signalAll();
        this.mAuthInfoLock.unlock();
    }

    @ReactMethod
    public void startAuthDownload(String vidId, String authStr, String format, String quality, boolean encrypted) {
        AliyunPlayAuth.AliyunPlayAuthBuilder playAuthBuilder = new AliyunPlayAuth.AliyunPlayAuthBuilder();
        playAuthBuilder.setVid(vidId);
        playAuthBuilder.setPlayAuth(authStr);
        playAuthBuilder.setFormat(format);
        playAuthBuilder.setQuality(quality);
        playAuthBuilder.setIsEncripted(encrypted? 1:0);
        AliyunPlayAuth playAuth = playAuthBuilder.build();

        AliyunDownloadMediaInfo info = new AliyunDownloadMediaInfo();
        info.setVid(vidId);
        info.setFormat(format);
        info.setQuality(quality);
        info.setEncripted(encrypted? 1:0);
        info.setSavePath(mStoragePath);

        mDownloadManager.prepareDownloadMedia(playAuth);
        mDownloadManager.addDownloadMedia(info);
        mDownloadManager.startDownloadMedia(info);
    }
}

File: ios/AliyunPlayView/AliyunVodDownloadModule.m
//
//  AliyunVodDownloadModule.m
//  AliyunPlayView
//
//  Created by vun on 2020/1/11.
//

#import "AliyunVodDownloadModule.h"

@implementation AliyunVodDownloadModule {
    Boolean mAuthInfoSet;
    NSCondition *mAuthInfoLock;
    NSString *mAuthInfoVidId;
    NSString *mAuthInfoAuthStr;
    NSString *mSavePath;
    NSMutableDictionary *mCurrentDownloadMediaInfo;
    AliMediaDownloader *mDownloader;
}

NSString *const EVENT_UPDATE_AUTH = @"AliyunVod.Downloader.UpdateAuth";
NSString *const EVENT_PREPARED = @"AliyunVod.Downloader.Prepared";
NSString *const ALI_DL_EVENT_START = @"AliyunVod.Downloader.Start";
NSString *const EVENT_PROGRESS = @"AliyunVod.Downloader.Progress";
NSString *const EVENT_COMPLETE = @"AliyunVod.Downloader.Completed";
NSString *const EVENT_STOP = @"AliyunVod.Downloader.Stop";
NSString *const EVENT_ERROR = @"AliyunVod.Downloader.Error";

RCT_EXPORT_MODULE()

- (instancetype)init
{
    self = [super init];
    if (self) {
        mAuthInfoLock = [[NSCondition alloc] init];
        mAuthInfoSet = NO;
        
        mSavePath = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
        mCurrentDownloadMediaInfo = [[NSMutableDictionary alloc] init];
    }
    return self;
}

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

-(void) initalize {
    NSLog(@"ios initialize");
    
    mAuthInfoLock = [[NSCondition alloc] init];
    mAuthInfoSet = NO;
}

- (NSArray<NSString *> *)supportedEvents {
    return @[EVENT_UPDATE_AUTH, EVENT_STOP, ALI_DL_EVENT_START, EVENT_ERROR, EVENT_COMPLETE, EVENT_PREPARED, EVENT_PROGRESS];
}

-(NSString*) onGetPlayAuth:(NSString*)vid format:(NSString*)format quality:(AliyunVodPlayerVideoQuality)quality {
    [self sendEventWithName:EVENT_UPDATE_AUTH body:@{@"vidId": vid}];
    [mAuthInfoLock lock];
    while(!mAuthInfoSet) {
        [mAuthInfoLock wait];
    }
    
    return mAuthInfoAuthStr;
}

RCT_EXPORT_METHOD(setAuth:(NSString*)vidId authStr:(NSString*)authStr) {
    [mAuthInfoLock lock];
    
    mAuthInfoVidId = vidId;
    mAuthInfoAuthStr = authStr;
    mAuthInfoSet = YES;
    
    [mAuthInfoLock signal];
    [mAuthInfoLock unlock];
}

-(void) onPrepare:(NSArray<AliyunDownloadMediaInfo*>*)mediaInfos {
    [self sendEventWithName:EVENT_PREPARED body:mediaInfos];
}

-(void)onPrepared:(AliMediaDownloader *)downloader mediaInfo:(AVPMediaInfo *)info {
    NSArray<AVPTrackInfo*>* tracks = info.tracks;
    [downloader selectTrack:[tracks objectAtIndex:0].trackIndex];
    [downloader start];
    
    [self sendEventWithName:EVENT_PREPARED body:@{@"@items": @[info]}];
}

-(void) onStart:(AliyunDownloadMediaInfo*)mediaInfo {
    [self sendEventWithName:ALI_DL_EVENT_START body:@{@"@items": @[mediaInfo]}];
}

-(void) onProgress:(AliyunDownloadMediaInfo*)mediaInfo {
    [self sendEventWithName:EVENT_PROGRESS body:@[mediaInfo]];
}

-(void)onDownloadingProgress:(AliMediaDownloader *)downloader percentage:(int)percent {
    NSNumber *x = [NSNumber numberWithFloat:((float)percent/100.0)];
    
    [mCurrentDownloadMediaInfo setValue:x forKey:@"progress"];
    [mCurrentDownloadMediaInfo setValue:downloader.downloadedFilePath forKey:@"savePath"];
    [self sendEventWithName:EVENT_PROGRESS body:@{@"media":@{@"@items": @[mCurrentDownloadMediaInfo]}, @"progress": x}];
}

-(void) onStop:(AliyunDownloadMediaInfo*)mediaInfo {
    [self sendEventWithName:EVENT_STOP body:@[mediaInfo]];
}

-(void)onCompletion:(AliMediaDownloader *)downloader {

    
    [mCurrentDownloadMediaInfo setValue:downloader.downloadedFilePath forKey:@"savePath"];
    
    [mDownloader destroy];
    downloader = nil;
    mDownloader = nil;
    
    [self sendEventWithName:EVENT_COMPLETE body:@{@"@items": @[mCurrentDownloadMediaInfo]}];
}

-(void)onError:(AliMediaDownloader *)downloader errorModel:(AVPErrorModel *)errorModel {
    [self sendEventWithName:EVENT_ERROR body:@{
        @"media": mCurrentDownloadMediaInfo,
        @"i": [NSString stringWithFormat:@"%lu", errorModel.code],
        @"s1": errorModel.message
    }];
}

RCT_EXPORT_METHOD(startAuthDownload:(NSString*)vidId authStr:(NSString*)authStr format:(NSString*)format quality:(NSString*)quality encrypted:(BOOL*)encrypted) {
    AVPVidAuthSource* authSrc = [[AVPVidAuthSource alloc] init];
    [authSrc setVid:vidId];
    [authSrc setPlayAuth:authStr];
    [authSrc setQuality:quality];
    [authSrc setRegion:@"cn-shanghai"];
    
    authSrc.vid = vidId;
    
    mDownloader = [[AliMediaDownloader alloc] init];
    
    NSString *savePath = [NSString stringWithFormat:@"%@/%@.%@", mSavePath, vidId, format];
    NSLog(savePath);
    
    [mCurrentDownloadMediaInfo setValue:vidId forKey:@"vid"];
    [mCurrentDownloadMediaInfo setValue:0 forKey:@"progress"];
    [mCurrentDownloadMediaInfo setValue:savePath forKey:@"savePath"];
    [mCurrentDownloadMediaInfo setValue:format forKey:format];
    
    [mDownloader setDelegate:self];
    [mDownloader setSaveDirectory:mSavePath];
    [mDownloader prepareWithPlayAuth:authSrc];
}

@end
